src/antidote_crdt_fat_counter.erl

```diff
@@ -0,0 +1,159 @@
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% antidote_crdt_fat_counter: A convergent, replicated, operation based Fat Counter
+%% The state of this fat counter is list of pairs where each pair is an integer
+%% and a related token.
+%% Basically when the counter recieves {increment, N} or {decrement, N} it generates
+%% a pair {N, NewToken}.
+%% On update, all seen tokens are removed and the new pair is then added to the state.
+%% This token keeps growing ("Fat" Counter) but it useful as it allows the reset
+%% functionaility, On reset(), all seen tokens are removed.
+
+-module(antidote_crdt_fat_counter).
+
+-behaviour(antidote_crdt).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([ new/0,
+          value/1,
+          downstream/2,
+          update/2,
+          equal/2,
+          to_binary/1,
+          from_binary/1,
+          is_operation/1,
+          is_bottom/1,
+          require_state_downstream/1
+        ]).
+
+-type uniqueToken() :: term().
+-type state() :: [{integer(), uniqueToken()}].
+-type op() ::
+    {increment, integer()}
+    | {decrement, integer()}
+    | {reset, {}}.
+-type effect() ::
+      {integer(), uniqueToken()}
+      | [uniqueToken()].
+
+%% @doc Create a new, empty fat counter
+-spec new() -> state().
+new() ->
+    [].
+
+%% @doc The value of this counter is equal to the sum of all the values
+%% having tokens.
+-spec value(state()) -> integer().
+value(FatCounter) ->
+    lists:sum([V || {V, _} <- FatCounter]).
+
+
+-spec downstream(op(), state()) -> {ok, effect()}.
+downstream(Op, FatCtr) ->
+    Token = unique(),
+    case Op of
+        {increment, Value} when is_integer(Value) ->
+            {ok, {Value, Token}};
+        {decrement, Value} when is_integer(Value) ->
+            {ok, {-Value, Token}};
+        {reset, {}} ->
+            Overridden = [Tok || {_, Tok} <- FatCtr],
+            {ok, lists:sort(Overridden)}
+    end.
+
+-spec unique() -> uniqueToken().
+unique() ->
+    crypto:strong_rand_bytes(20).
+
+
+-spec update(effect(), state()) -> {ok, state()}.
+update({Value, Token}, FatCtr) ->
+    % insert new value
+    {ok, FatCtr ++ [{Value, Token}]};
+update(Overridden, FatCtr) ->
+  SortedFatCtr = lists:sort(fun({_, A}, {_, B}) -> A =< B end, FatCtr),
+  FatCtr2 = [{V, T} || {V, T} <- SortedFatCtr, not lists:member(T, Overridden)],
+  {ok, FatCtr2}.
```

Comment:
Here you should decide: Either you use sorted state, so that you can do the update here in linear time (like in the set-implementations), or you just take the simple way without sorting. Currently you are sorting, but I don't see any benefit from sorting.

If you choose sorting, I would always keep the state sorted and use an `orddict` or `ordset` for the state to make this clear.
If you use an `ordset` for the state and change the downstream of `reset` to be the complete state, you could probably use `ordsets:subtract` here.